lib/docker.ts

```diff
@@ -71,19 +91,48 @@ class Docker {
   }
 
   public async catSafe(filename: string) {
-    try {
-      return await this.run("cat", [filename]);
-    } catch (error) {
-      const stderr: string = error.stderr;
-      if (typeof stderr === "string") {
-        if (
-          stderr.indexOf("No such file") >= 0 ||
-          stderr.indexOf("file not found") >= 0
-        ) {
-          return { stdout: "", stderr: "" };
+    return await this.runSafe("cat", [filename]);
+  }
+
+  public async lsSafe(path: string, recursive?: boolean) {
+    let params = "-1ap";
+    if (recursive) {
+      params += "R";
+    }
+    return await this.runSafe("ls", [params, path]);
+  }
+
+  /**
+   * Find files on a docker image according to a given list of glob expressions.
+   */
+  public async findGlobs(
+    globs: string[],
+    exclusionGlobs: string[] = [],
+    path: string = "/",
+    recursive: boolean = true,
+  ) {
+    const res: string[] = [];
+    const root = await this.lsSafe(path, recursive).then((output) =>
+      lsu.parseLsOutput(output.stdout),
+    );
+
+    lsu.iterateFiles(root, (f) => {
+      const filepath = fspath.join(f.path, f.name);
+      let exclude = false;
+      exclusionGlobs.forEach((g) => {
```

Comment:
Consider using a for loop and break when exclude is set to true

Reply:
I would prefer a function here. Normally you are right - we should break out of a loop as soon as possible. In this case the looping effort is insignificant. We'll only expect a few exclude patterns (<10) and the expensive minimatch is no longer executed once the exclude flag has been set. I prefer iterators to loops in JavaScript as variables have function score and not block scope - as such iterator functions are an effective way to avoid polluting the variable scope of the enclosing function with loop variables.